src/torchmetrics/functional/nominal/thiels_u.py

```diff
@@ -0,0 +1,155 @@
+        S(x|y): float
+
+    Example:
+        >>> from torchmetrics.functional import conditional_entropy
+        >>> _ = torch.manual_seed(42)
+        >>> preds = torch.randint(10, (10,))
+        >>> target = torch.randint(10, (10,))
+        >>> conditional_entropy(preds, target)
+        tensor(0.2773)
+    """
+    # reduce dim if ndim == 2
+    preds = preds.argmax(1) if preds.ndim == 2 else preds
+    target = target.argmax(1) if target.ndim == 2 else target
+    # handle the nan in data
+    preds, target = _handle_nan_in_data(preds, target, nan_strategy, nan_replace_value)
+
+    # set x and y
+    x, y = preds.numpy(), target.numpy()
+
+    y_counter = Counter(y)
+    xy_counter = Counter(list(zip(x, y)))
+    total_occurrences = sum(y_counter.values())
+    entropy = 0.0
+    for xy in xy_counter.keys():
+        p_xy = xy_counter[xy] / total_occurrences
+        p_y = y_counter[xy[1]] / total_occurrences
+        entropy += p_xy * math.log(p_y / p_xy, log_base)
+    return torch.tensor(entropy)
+
+
+def theils_u(
+    preds: Tensor,
+    target: Tensor,
+    nan_strategy: Literal["replace", "drop"] = "replace",
+    nan_replace_value: Optional[Union[int, float]] = 0.0,
+    _PRECISION: float = 1e-13,
+) -> Tensor:
+    r"""Computes Theil's U Statistic (Uncertainty Coefficient). The value is between 0 and 1, i.e. 0 means y has no
+    information about x while value 1 means y has complete information about x.
+
+    The implementation is based on the code provided in https://github.com/shakedzy/dython/
+
+    Args:
+        preds: 1D or 2D tensor of categorical (nominal) data
+            - 1D shape: (batch_size,)
+            - 2D shape: (batch_size, num_classes)
+        target: 1D or 2D tensor of categorical (nominal) data
+            - 1D shape: (batch_size,)
+            - 2D shape: (batch_size, num_classes)
+        nan_strategy: Indication of whether to replace or drop ``NaN`` values
+        nan_replace_value: Value to replace ``NaN``s when ``nan_strategy = 'replace'``
+        _PRECISION: used to round off values above 1 + _PRECISION and below 0 - _PRECISION
+    Returns:
+        Thiel's U Statistic: Tensor
+
+    Example:
+        >>> from torchmetrics.functional import theils_u
+        >>> _ = torch.manual_seed(42)
+        >>> preds = torch.randint(10, (10,))
+        >>> target = torch.randint(10, (10,))
+        >>> theils_u(preds, target)
+        tensor(0.8530)
+    """
+    # reduce dim if ndim == 2
+    preds = preds.argmax(1) if preds.ndim == 2 else preds
+    target = target.argmax(1) if target.ndim == 2 else target
+    # handle the nan in data
+    preds, target = _handle_nan_in_data(preds, target, nan_strategy, nan_replace_value)
+
+    # compute conditional entropy
+    s_xy = conditional_entropy(preds, target)
+
+    x_counter = Counter(preds.numpy())
+    total_occurrences = sum(x_counter.values())
+    p_x = list(map(lambda n: n / total_occurrences, x_counter.values()))
+    s_x = ss.entropy(p_x)
```

Comment:
Believe we should be able to implement this without a dependency on `scipy` :]

Reply:
I updated it without any package dependence. Can you check it out?

Reply:
Yeah, fixed this, now using the tensor based `confmat` based solution